### PR TITLE
Adds automatic sample weighting

### PR DESCRIPTION
**Pull request by danielhertenstein**
This PR adds a sample weighting factor to balance the importance of positive samples with negative samples. See: https://pytorch.org/docs/stable/nn.html#bcewithlogitsloss

The main question is will we always want this? I think yes as a default, but it can always be made optional at a later date if need be.

### CODE REVIEW

**Comment by erikrose** [comment]
Hmm, this change makes at least one set of vectors in lockwise-proof-of-concept give worse accuracy (not that accuracy is the be-all and end-all, but it makes it an unclear win).

Old trainer:

    % fathom-train lockwise-proof-of-concept/vectors/trainingNextRegisterAttrsOrContent.json -i 500 -s -a lockwise-proof-of-concept/vectors/validationNextRegisterAttrsOrContent.json  -l .1
    Stopping early at iteration 497, just before validation error rose.
    {"coeffs": [
            ["nextAnchorIsJavaScript", 1.1796753406524658],
            ["nextButtonTypeSubmit", 4.6356520652771],
            ["nextInputTypeSubmit", 4.415358543395996],
            ["nextInputTypeImage", 6.90437650680542],
            ["nextLoginAttrs", 0.07329149544239044],
            ["nextButtonContentContainsLogIn", -0.5901219248771667],
            ["nextButtonContentIsLogIn", 4.880642890930176],
            ["nextInputContentContainsLogIn", 4.2968621253967285],
            ["nextInputContentIsLogIn", 4.123159885406494],
            ["nextButtonContentIsNext", -0.06832335889339447],
            ["nextNearUsername", 4.531530857086182],
            ["nextRegisterAttrsOrContent", -3.3321805000305176]
        ],
     "bias": -8.670398712158203}
      Training accuracy per tag:  0.98312    95% CI: (0.97153, 0.99472)
                             FP:  0.00000    95% CI: (0.00000, 0.00000)
                             FN:  0.01688    95% CI: (0.00470, 0.02905)
                      Precision:  1.00000    Recall: 0.81818

    Validation accuracy per tag:  0.97347    95% CI: (0.95924, 0.98770)
                             FP:  0.00000    95% CI: (0.00000, 0.00000)
                             FN:  0.02653    95% CI: (0.01139, 0.04167)
                      Precision:  1.00000    Recall: 0.77193

      Training accuracy per page: 1.00000    95% CI: (1.00000, 1.00000)
    Validation accuracy per page: 0.96875    95% CI: (0.92612, 1.00000)

This new one, with pos_weight:

    % fathom-train lockwise-proof-of-concept/vectors/trainingNextRegisterAttrsOrContent.json -i 500 -s -a lockwise-proof-of-concept/vectors/validationNextRegisterAttrsOrContent.json  -l .1
    Stopping early at iteration 111, just before validation error rose.
    {"coeffs": [
            ["nextAnchorIsJavaScript", 1.0658918619155884],
            ["nextButtonTypeSubmit", 3.323981761932373],
            ["nextInputTypeSubmit", 3.46545672416687],
            ["nextInputTypeImage", 5.817039489746094],
            ["nextLoginAttrs", -0.051404666155576706],
            ["nextButtonContentContainsLogIn", 1.0390106439590454],
            ["nextButtonContentIsLogIn", 2.5551247596740723],
            ["nextInputContentContainsLogIn", 2.8567118644714355],
            ["nextInputContentIsLogIn", 2.6008312702178955],
            ["nextButtonContentIsNext", 0.16415071487426758],
            ["nextNearUsername", 3.5782880783081055],
            ["nextRegisterAttrsOrContent", -2.4899771213531494]
        ],
     "bias": -5.3140740394592285}
      Training accuracy per tag:  0.96414    95% CI: (0.94739, 0.98088)
                             FP:  0.03586    95% CI: (0.00000, 0.09081)
                             FN:  0.00000    95% CI: (0.00000, 0.00000)
                      Precision:  0.72131    Recall: 1.00000

    Validation accuracy per tag:  0.96327    95% CI: (0.94661, 0.97992)
                             FP:  0.02653    95% CI: (0.00000, 0.06825)
                             FN:  0.01020    95% CI: (0.00074, 0.01967)
                      Precision:  0.80000    Recall: 0.91228

      Training accuracy per page: 0.91667    95% CI: (0.84673, 0.98660)
    Validation accuracy per page: 0.89062    95% CI: (0.81416, 0.96709)

It seems like the main difference is that this new trainer does a lot of this for the training set:

     failure  on  fp15.html. Confidence: 0.67910939 There were no right choices, but highest-scorer had high confidence anyway.

Consequently, there are several new instances of the same error in the validation set. There is also one new validation error where it outright picks the wrong element though a right one is available.

This does crank up recall at the expense of precision, which is a nice knob to be able to turn. Maybe this would do better as an option, off by default until we find a knob setting that beats the old trainer on accuracy most of the time. Perhaps we should even let the user turn the knob from the commandline. I'm very open to a longer, more realtime conversation about this whole thing.

Caveats:
* I tried only one set of vectors.
* The features in that ruleset may have been selected to drive the old trainer to highest accuracy.

**Comment by danielhertenstein** [comment]
Hmm, okay. I can go the route of making it a command line argument. It does seem best suited as a knob to tune precision vs. recall, and priorities of our rulesets will likely differ.
